### PR TITLE
[12.0][IMP] hr_timesheet_sheet: different validation models

### PR DESCRIPTION
**Pull request by alexey-pelykh**
This module adds option to configure how the timesheet sheet validation is done. Reviewer is
configurable to be HR Officer/Manager, Department Manager, Direct Manager, or
Project Manager.

To change who reviews submitted sheets, go to *Configuration > Settings > Timesheet Options*
and configure **Timesheet Sheet Review Policy** accordingly.

### CODE REVIEW

**Comment by alexey-pelykh** [comment]
@mreficent @astirpe it would be good to hear out your thoughts on this WIP improvement, especially about
```
        if not self.env.user.has_group('hr_timesheet.group_hr_timesheet_user'):
            raise UserError(
                _('Only an HR Officer or Manager can approve sheets.'))
```
being changed to check for `hr.group_hr_user` as error message sounds right while logic isn't.

**Comment by MiquelRForgeFlow** [comment]
I think it should be better if you do a separate module named `hr_timesheet_sheet_review_policy` where you add all this `review_policy` related thing, maybe. I would appreciate that very much.

**Comment by alexey-pelykh** [comment]
@mreficent good point to think about. What pros/cons you have in mind?

**Comment by alexey-pelykh** [comment]
@mreficent @astirpe from code that's right now in master, it seems that anyone can approve anyone (as long as they have access to timesheets), is that by design?

**Comment by alexey-pelykh** [comment]
@mreficent @astirpe this one is ready for review, hopefully you've seen my email about call for discussion

**Comment by MiquelRForgeFlow** [comment]
If you don't want to separate the refactoring from the Review Policy thing, then at least do in separate commits. I don't see why all your changes should be compacted in just one commit.

**Comment by alexey-pelykh** [comment]
@mreficent I found that decoupling that into separate module would be quite an effort - anyhow "default" review policy would have been implemented in `hr_timesheet_sheet`, so I decided that it would be better to implement most useful policies in this module, using extra 15 lines of code, than extract them into a separate one.

Regarding separating the commits - I added few enhancements to make module more extendable, e.g. extracted couple of methods. That's a pure "refactor"/"more hooks" commit - you mean extract such one?

**Comment by MiquelRForgeFlow** [comment]
Yeah, one commit for this "refactor/more hooks", other for the addition of review policy, at least. With separate commits, it's easier to review that having all together.

**Comment by alexey-pelykh** [comment]
@mreficent same PR or separate one?

**Comment by MiquelRForgeFlow** [comment]
I would prefer separate PR, but I do not want to bother too much :S

**Comment by alexey-pelykh** [comment]
https://github.com/OCA/timesheet/pull/264

**Comment by astirpe** [comment]
I need some time to check this PR and the other. I hope I can do it this week. Thanks!

**Comment by alexey-pelykh** [comment]
@astirpe thanks for your time! 
@mreficent this has been updated and rebased

**Comment by alexey-pelykh** [comment]
@mreficent @astirpe 🙏 if I can grab some of your precious time

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 1]
```diff
@@ -164,6 +199,59 @@ def _default_employee(self):
...
+        """ Hook for extensions """
+        review_policy = self.env.user.company_id.timesheet_sheet_review_policy
+        field = None
+        if review_policy == 'project_manager' and view_type == 'tree':
+            field = view.find(".//field[@name='project_id']")
+        elif review_policy == 'department_manager' and view_type == 'tree':
+            field = view.find(".//field[@name='department_id']")
+
+        if field is not None:
+            field.set('invisible', '0')
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:249]: Shouldn't all this black magic be moved to a separate custom module? I see it's made just for the visibility of two fields in the tree view. I think it doesn't worth it.

**Review by astirpe — CHANGES_REQUESTED** [review]
I think the most of the proposed changes should be moved to a separate module, as the approval process of a timesheet sheet can vary a lot among different companies. Trying to find a generic way for all the kind of companies seems to me something debatable.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:249]: But it's the only way to make those columns conditionally invisible, and leaving them as is would only pollute the UI

**Comment by alexey-pelykh** [comment]
@astirpe this was brought up by @mreficent earlier, the actual validation models can be extracted to a separate module, but the hooking points and related refactor has to be done within this module. In other words, there are approximately 20 lines of code that implement different models, all other are needed to allow those models to act. On top of that - this module is extendable, and I've decided to implement most commonly-used models within it.

**Comment by alexey-pelykh** [comment]
In other words, if specific company has group-based validation model or something - it's possible to make an extension to this module that would have approximately 10-20 lines of code in it that actually implement one extra mode of validation.

**Comment by MiquelRForgeFlow** [comment]
I would call the new module as "`hr_timesheet_sheet_review_policy`" or "`hr_timesheet_sheet_approval_policy`"

**Comment by alexey-pelykh** [comment]
@mreficent to make that module possible, `hr_timesheet_sheet` module still has to be changed, to have concept of customizable validation models. Due to this fact, I don't understand value of ejecting commonly-used policies to separate module

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:249]: Module `hr_timesheet_sheet` should be a base module for being extended (and thanks for splitting the methods to hook methods) but this kind of changes are going a bit too far IMO. I would prefer to keep the module as a basis, although it became already quite complex compared to the original V10 version.

Regarding this particular part of code, are you sure that the conditionally invisible fields are necessary? The original tree view doesn't have those two fields and therefore it's open for being customized.

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:249]: Well, those fields were added for `department_manager` validation policy, so for all others these fields are not valuable to be seen

Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:249]: As I said, I think it doesn't worth it.

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:249]: Does not worth adding `department_manager` policy? Does not worth hiding policy-specific columns?

Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:249]: Not by adding such black magic in this module.

I would love to continue to discuss about this, but my holidays are now starting.

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:249]: :) It's not black magic, it's even used in Odoo itself :P Please ping me when you're back from holidays

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 2]
```diff
@@ -164,6 +199,59 @@ def _default_employee(self):
...
+        self._modify_view_for_policy(view_type, view)
+        res['arch'] = etree.tostring(
+            view,
+            encoding='unicode',
+        ).replace('\t', '')
+
+        return res
+
+    @api.model
+    def _modify_view_for_policy(self, view_type, view):
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:239]: @mreficent @astirpe to give you better insights: this method has 4 lines of code specific to validation models.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 3]
```diff
@@ -210,28 +336,49 @@ def _check_start_end_dates(self):
                     _('The start date cannot be later than the end date.'))
 
     @api.multi
-    def _get_overlapping_sheet_domain(self, user_id=False):
+    def _get_complete_name_components(self):
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:339]: @mreficent @astirpe to give you better insights: this method has another 4 lines of code specific to validation models.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 4]
```diff
@@ -210,28 +336,49 @@ def _check_start_end_dates(self):
...
+        self.ensure_one()
+        result = [self.employee_id.name_get()[0][1]]
+        if self.review_policy == 'department_manager':
+            result += [self.department_id.name_get()[0][1]]
+        elif self.review_policy == 'project_manager':
+            result += [self.project_id.name_get()[0][1]]
+        return result
+
+    @api.multi
+    def _get_overlapping_sheet_domain(self):
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:350]: @mreficent @astirpe to give you better insights: this method has another 4 lines of code specific to validation models.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 5]
```diff
@@ -273,6 +430,21 @@ def _check_company_id_add_line_task_id(self):
                     _('The Company in the Timesheet Sheet and in '
                       'the Task must be the same.'))
 
+    @api.multi
+    def _get_possible_reviewers(self):
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:434]: @mreficent @astirpe to give you better insights: this method has another 6 lines of code specific to validation models.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 6]
```diff
@@ -498,21 +695,55 @@ def action_timesheet_confirm(self):
...
+        if not reviewer:
+            raise UserError(_(
+                'In order to review a timesheet sheet, your user needs to be'
+                ' linked to an employee.'
+            ))
+        return reviewer
+
+    @api.multi
+    def _check_can_review(self):
+        for sheet in self.filtered(lambda x: not x.can_review):
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:724]: @mreficent @astirpe to give you better insights: this method has another 12 lines of code specific to validation models.

**Comment by alexey-pelykh** [comment]
@mreficent @astirpe I walked over the code of the PR, there are approx. 30 lines of code that are related to other validation models apart of default one. Do you honestly think it worth creating 3 extra models, one per each validation model, to add approx. 10 lines of code per each module?

**Comment by alexey-pelykh** [comment]
@tarteo if you have a spare minute, it would be great to get your review as well

**Comment by alexey-pelykh** [comment]
@CasVissers maybe you're interested in reviewing this?

**Comment by alexey-pelykh** [comment]
Closed in favor of #272

**Pull request closed without merging**